### ISSUE TITLE
get_input_window_for_output_time() and predict() need to adjust for bin_closed and bin_label options

### ISSUE THREAD

**Issue by stephen-frank** [NREL/Wattile#253]
Model configuration includes `bin_closed` and `bin_label` options (under `data_processing` > `resample`), which by default are:

- `bin_closed = "right"`
- `bin_label = "right"`

With the defaults, the `get_input_window_for_output_time()` and `predict()` methods act as expected:

1. When predicting a single timestamp, the correct window of time gets selected
2. When predicting a single timestamp, the model output has the expected nominal timestamp and horizon

With other cases, either `get_input_window_for_output_time()` fails to provide a time window that is what the model actually needs, or the output of `predict` has nominal timestamp offset from the expected value.

Results from test models that @JanghyunJK supplied are below...

### `bin_closed = "right"`, `bin_label = "right"`
Works as expected:

![image](https://user-images.githubusercontent.com/1534082/207970246-76f9ee3e-6ef7-4bb4-86b1-90de765e094b.png)

### `bin_closed = "right"`, `bin_label = "left"`
Returns results, but the nominal time is 15 minutes earlier than expected. In this case, `get_input_window_for_output_time()` was called using 8:00 AM and the prediction output has a nominal timestamp of 7:45 AM.

![image](https://user-images.githubusercontent.com/1534082/207970563-a647ffd4-c069-4e5d-90f1-a329eb99a894.png)

Here is what `get_input_window_for_output_time()` is returning:

```
session.pyEval("prediction_time") >> 15-Dec-2022 Thu 1:00:00PM UTC
session.pyEval("model.get_input_window_for_output_time(prediction_time)") >> (Timestamp('2022-12-15 07:00:00+0000', tz='UTC'), Timestamp('2022-12-15 13:00:00+0000', tz='UTC'))
```

### `bin_closed = "left"`, `bin_label = "right"`
`predict()` returns an error:

```
axon::EvalErr: Func failed: pyEval(PySession py,Str stmt); args: (PyMgrSession,Str)
  sys::IOErr: Python failed: num_samples should be a positive integer value, but got num_samples=0
Traceback (most recent call last):
  File "/usr/src/app/hxpy/hxpy.py", line 67, in run
    self._exec(instr, local_vars)
  File "/usr/src/app/hxpy/hxpy.py", line 95, in _exec
    return exec(code, local_vars, local_vars)
  File "<string>", line 1, in <module>
  File "/wattile/wattile/models/AlgoMainRNNBase.py", line 187, in predict
    val_loader = self.to_data_loader(
  File "/wattile/wattile/models/alfa_model.py", line 47, in to_data_loader
    return DataLoader(
  File "/usr/local/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 353, in __init__
    sampler = RandomSampler(dataset, generator=generator)  # type: ignore[arg-type]
  File "/usr/local/lib/python3.9/site-packages/torch/utils/data/sampler.py", line 107, in __init__
    raise ValueError("num_samples should be a positive integer "
ValueError: num_samples should be a positive integer value, but got num_samples=0
 [proj_wattile_test::wattilePythonModelPredict:140]
=== Axon Trace ===
  wattilePythonModelPredict (proj_wattile_test::wattilePythonModelPredict:140)
  wattilePythonTask (proj_wattile_test::wattilePythonTask:29)
  anon-1 (taskExpr:1)
```

The relevant line here is: `num_samples should be a positive integer value, but got num_samples=0`

I can trace it back to an empty `val_df`:

```
session.pyExec("_, val_df = prep_for_rnn(configs, predictor_data_frame)")
return session.pyEval("val_df.to_string")
```

...which yields:

```
<bound method DataFrame.to_string of Empty DataFrame
Columns: [sin_HOD_min_lag24, cos_HOD_min_lag24, ...]
Index: []

[0 rows x 4951 columns]>
```

In other words, when `bin_closed = "left"` and I pass it the window of data that `get_input_window_for_output_time()` specifies, then the data processing pipeline returns an empty data frame.

I did try shifting the entire read span left or right by 15 minutes, and still come up with the empty data frame. (It is worth noting that I also have `start_time` and `end_time` in the configs set to match the window that `get_input_window_for_output_time()` returned... not sure if this makes a difference but it seems like it might.)

### `bin_closed = "left"`, `bin_label = "left"`
Same error behavior as `bin_closed = "left"`, `bin_label = "right"`

**Comment by stephen-frank** [comment]
After fixing, I recommend setting up unit tests for all four cases which include:

- Predicting only for a single timestamp
- Ensuring input data windowed per `get_input_window_for_output_time()` successfully yields exactly one prediction
- Setting `data_start` and `data_end` to match the window returned by `get_input_window_for_output_time()` as well, to ensure filtering doesn't break things
- Ensuring the nominal timestamp of the prediction matches `get_input_window_for_output_time()`

**Comment by stephen-frank** [comment]
Oh, I forgot to mention: the reason any of this matters is that SkySpark's default (and only) behavior for rollups is left-closed, left-labeled. So, if you roll up all data from 8:00:00 to 8:15:00...

- Input data with timestamp `8:00:00` is _included_
- Input data with timestamp `8:15:00` is _excluded_
- The output data point has timestamp `8:00:00`

This is basically opposite of Python's default behavior, which is why we need `bin_closed` and `bin_label` options.

**Comment by stephen-frank** [comment]
Widening `input_start` and `input_end` by 1 hour did not solve the problem with `bin_closed="left"` returning empty `val_df` from `prep_for_rnn`:

```
.pyExec("input_start = model.get_input_window_for_output_time(first_prediction_time)[0]")
.pyExec("input_end = model.get_input_window_for_output_time(last_prediction_time)[1]")
...

.pyExec("configs['data_input']['start_time'] = (input_start - pandas.Timedelta('1 hours')).isoformat()")
.pyExec("configs['data_input']['end_time'] = (input_end + pandas.Timedelta('1 hours')).isoformat()")
```

**Comment by stephen-frank** [comment]
SkySpark code for reference: https://github.com/NREL/nrelWattileExt/blob/wattile-zero-point-one-changes/lib/pythonFuncs.trio#L57-L228

**Comment by JanghyunJK** [comment]
Thus far, I think there are three things happening:

- When developing this `get_input_window_for_output_time` method for prediction, I think I was incorrectly slicing the data for testing. So, I'm implementing this below for further testings.

  - left-closed --> `data.loc[(input_start<=data.index)&(data.index<input_end),:]`
  - right-closed --> `data.loc[(input_start<data.index)&(data.index<=input_end),:]`

- Because the basis of slicing of data was done incorrectly, I think `window_offset` parameter in `get_input_window_for_output_time` also needs revision.

  - from `pd.Timedelta(lag_interval) * lag_count` to `pd.Timedelta(lag_interval) * (lag_count+1)`

- I haven't found a good reasoning for this last one. We go through `resample` and `rolling` methods when creating rolling statistics features. While both methods include an argument called `closed`, they don't behave the same; `resample` opens/closes exactly at the timestamp and `rolling` opens/closes for the first (left) or last (right) sample. So, now I'm leaning towards hard-coding `rolling` method's `closed` argument to `right`; `resample` method's `closed` argument will be controlled by our configs parameter `bin_closed`. 

All things mentioned above are reflected in my [working notebook](https://github.com/NREL/Wattile/blob/notebook-revision-edge-consideration/notebooks/exploratory/data_processing_edge_testing.ipynb).

And once I get confirmation for the [question I posted on Teams](https://teams.microsoft.com/l/message/19:caffaa1617874d3aab9ef73c903753f4@thread.skype/1672167296533?tenantId=a0f29d7e-28cd-4f54-8442-7885aee7c080&groupId=a673ac7c-234b-4f92-9636-08629146bfd2&parentMessageId=1672167296533&teamName=Intelligent%20Campus%20Team&channelName=Predictive%20Analytics&createdTime=1672167296533&allowXTenantAccess=false), I'll be able to create a PR or make more changes and then create PR.

**Comment by stephen-frank** [comment]
Ah! I recall now there were originally cases `rolling` and `binned`, with `binned` being what was supposed to be tested above (I think?). However, if I look now at the [Configs README](https://github.com/NREL/Wattile/blob/main/docs/Configs.md), There is no longer a reference to `binned` and only rolling windows are discussed. So now I am wondering if there was a disconnect somewhere along the way during cleanup of the code.

**Comment by haneslinger** [comment]
here's the altered tests! https://github.com/NREL/Wattile/compare/Fix-test_get_input_window_for_output_time?expand=1

**Comment by stephen-frank** [comment]
Yes, I think those tests look right.

**Comment by stephen-frank** [comment]
I think we can close complete, yes?

**Comment by haneslinger** [comment]
yep!
